src/Microsoft.ServiceHub.Framework/AsyncNamedPipeClientStream.cs

```diff
@@ -0,0 +1,174 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.IO.Pipes;
+using System.Runtime.InteropServices;
+using System.Runtime.Versioning;
+using System.Security.Principal;
+using Microsoft.Win32.SafeHandles;
+
+namespace Microsoft.ServiceHub.Framework
+{
+	/// <summary>
+	/// Named pipe client that avoids TimeoutExceptions and burning CPU.
+	/// </summary>
+	[SupportedOSPlatform("windows")]
+	internal class AsyncNamedPipeClientStream : PipeStream
+	{
+		private const int SECURITYSQOSPRESENT = 0x00100000;
+		private const int GENERICREAD = unchecked((int)0x80000000);
+		private const int GENERICWRITE = 0x40000000;
```

Comment:
Can we replace all these constants and your `extern` methods with CsWin32? This project already uses it, so it should be a simple matter of adding these APIs to NativeMethods.txt.